### PR TITLE
Py3 unicode

### LINKED ISSUE #36031: Windows PY3 resolved issues outstanding for PY2, non-salt
### Description of Issue/Question

This is a list of issues which exist in python 2 which have been resolved in python  3.
Feel free to post any missing fixes.
- win reg api under PY2 is non-Unicode check a Unicode registry.  PY3 as full Unicode API support
- Fix islink and effectively os.walk   https://bugs.python.org/issue1578269
- Speed improvements to os.walk https://bugs.python.org/issue23605

### LINKED ISSUE #54447: SALT-MASTER PATCHED state.sls causes UnicodeDecodeError: 'utf8' codec can't decode byte 0xfc (this is ü)
### Description of Issue

A comment causes an exception but is not shown at all.

I would be extremely grateful if Salt would have an extremely robust way to display text, at least in the exception.  E.g. delete all non-ASCII characters.  

Character ü is not-ASCII but must not be a problem for Salt. 

### Setup
The comment is not reported.
I cannot tell which state in the sls causes the comment

### Steps to Reproduce Issue
salt MINION state.sls syncP2P

```
[ERROR   ] An un-handled exception was caught by salt's global exception handler:
UnicodeDecodeError: 'utf8' codec can't decode byte 0xfc in position 642: invalid start byte
Traceback (most recent call last):
  File "/bin/salt", line 10, in <module>
    salt_main()
  File "/usr/lib/python2.7/site-packages/salt/scripts.py", line 495, in salt_main
    client.run()
  File "/usr/lib/python2.7/site-packages/salt/cli/salt.py", line 197, in run
    self._output_ret(ret_, out, retcode=retcode)
  File "/usr/lib/python2.7/site-packages/salt/cli/salt.py", line 376, in _output_ret
    _retcode=retcode)
  File "/usr/lib/python2.7/site-packages/salt/output/__init__.py", line 96, in display_output
    display_data = try_printout(data, out, opts, **kwargs)
  File "/usr/lib/python2.7/site-packages/salt/output/__init__.py", line 43, in try_printout
    printout = get_printout(out, opts)(data, **kwargs)
  File "/usr/lib/python2.7/site-packages/salt/output/highstate.py", line 157, in output
    for host, hostdata in six.iteritems(data)
  File "/usr/lib/python2.7/site-packages/salt/output/highstate.py", line 335, in _format_host
    ret['comment'] = salt.utils.stringutils.to_unicode(ret['comment'])
  File "/usr/lib/python2.7/site-packages/salt/utils/stringutils.py", line 160, in to_unicode
    raise exc  # pylint: disable=raising-bad-type
UnicodeDecodeError: 'utf8' codec can't decode byte 0xfc in position 642: invalid start byte
```

### Versions Report
```
Salt Version:
           Salt: 2018.3.4

Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: Not Installed
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.7.2
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.5.6
   mysql-python: 1.2.5
      pycparser: Not Installed
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.5 (default, Jun 20 2019, 20:27:34)
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 15.3.0
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.1.4

System Versions:
           dist: centos 7.6.1810 Core
         locale: UTF-8
        machine: x86_64
        release: 3.10.0-957.5.1.el7.x86_64
         system: Linux
        version: CentOS Linux 7.6.1810 Core
```

Minion 2016.11.3

### PR DESCRIPTION
**Pull request by twangboy**
### What does this PR do?
Adds some tests for Unicode Support with Py3
Adds the ability to pass file content to the `@with_tempfile` helper
Adds the ability to pass the code page to use with run commands on Windows (`chcp`)

### What issues does this PR fix or reference?
Ref https://github.com/saltstack/salt/pull/56028
fixes https://github.com/saltstack/salt/issues/54447
fixes https://github.com/saltstack/salt/issues/36031

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**

- [x] Docs
- [x] Changelog - https://docs.saltstack.com/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
Yes

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

### CODE REVIEW

**Comment by markuskramerIgitt** [comment]
This is a step forward.

I replaced 437 with 65001 by notepad c:\salt\bin\Lib\site-packages\salt-3001-py3.7.egg\salt\modules\cmdmod.py

Observations:
- Run with salt-call on Windows, ` chcp 65001 > nul & ` does has no effect on the first run, but on the 2nd and 3rd.
- In a new Windows console, the problem reappears.
- Run with salt from the master, ` chcp 65001 > nul & ` has no effect on the first run, but on the 2nd and 3rd.

```
\salt\salt-call  cmd.run "dir \windows\temp\salt-test"
[ERROR   ] Failed to decode stdout from command chcp 65001 > nul & dir \windows\temp\salt-test, non-decodable characters have been replaced
local:
     Datentr�ger in Laufwerk C: ist Windows
     Volumeseriennummer: EA6E-94D5

     Verzeichnis von C:\windows\temp\salt-test

    10.07.2020  18:05    <DIR>          .
    10.07.2020  18:05    <DIR>          ..
    10.07.2020  18:05                12 ausgef�hrt.txt
    10.07.2020  17:51                13 Datentr�ger.txt
                   2 Datei(en),             25 Bytes
                   2 Verzeichnis(se), 328.308.600.832 Bytes frei

\salt\salt-call  cmd.run "dir \windows\temp\salt-test"
local:
     Datenträger in Laufwerk C: ist Windows
     Volumeseriennummer: EA6E-94D5

     Verzeichnis von C:\windows\temp\salt-test

    10.07.2020  18:05    <DIR>          .
    10.07.2020  18:05    <DIR>          ..
    10.07.2020  18:05                12 ausgeführt.txt
    10.07.2020  17:51                13 Datenträger.txt
                   2 Datei(en),             25 Bytes
                   2 Verzeichnis(se), 328.307.232.768 Bytes frei

```

## Further comment on PR 57515

### Removed chcp in cmdmod.py

For investigation, I installed a minion from git (`git checkout v3001` and `pip install -e .`)  and removed setting of chcp in cmdmod.py.

### Tried PYTHONUTF8

 [PEP540 UTF8-Mode in Python 3.7](https://www.python.org/dev/peps/pep-0540/) offers command line option `python.exe -X utf8` and environment variable PYTHONUTF8

As I don't know which salt-call is used, I used the environment variable.

Unfortunately PYTHONUTF8 has no effect :
- Setting $env:PYTHONUTF8 = 1  from PowerShell
- salt-call --local cmd.run "dir" from PowerShell
- ä is `„ `

### Tried chcp from PowerShell
chcp has instant effect, switches back and forth:

- Good
  - chcp 65001 from Powershell
  - salt-call --local cmd.run "dir" from PowerShell
  - ä is ä

- Bad
  - chcp 437 from Powershell
  - salt-call --local cmd.run "dir" from PowerShell
  - ä is `„ `

I think chcp should be called before python.exe.

**Review by cmcmarrow — APPROVED** [review]

File: salt/modules/cmdmod.py [hunk 1]
```diff
@@ -602,7 +603,7 @@ def _get_stripped(cmd):
         else:
             # On Windows set the codepage to US English.
             if python_shell:
```
Comment (cmcmarrow) [inline salt/modules/cmdmod.py:605]:
```suggestion
            if python_shell and chcp is not None:
                _run(cmd="chcp {0} > nul".format(chcp),
                        cwd=cwd,
                        stdin=stdin,
                        stdout=stdout,
                        stderr=stder,
                        output_encoding=output_encoding,
                        output_loglevel=output_loglevel=,
                        log_callback=log_callback,
                        runas=runas,
                        group=group,
                        shell=shell,
                       python_shell=python_shell,
                       env=env,
                       clean_env=clean_env,
                       prepend_path=prepend_path,
                       rstrip=rstrip,
                       template=template,
                       umask=umask,
                       timeout=timeout,
                       with_communicate=with_communicate,
                       reset_system_locale=reset_system_locale,
                       ignore_retcode=ignore_retcode,
                       saltenv=saltenv,
                       pillarenv=pillarenv,
                       pillar_override=pillar_override,
                       use_vt=use_vt,
                       password=password,
                       bg=bg,
                       encoded_cmd=encoded_cmd,
                       success_retcodes=success_retcodes,
                       chcp=None,
                       **kwargs)
```

Comment (cmcmarrow) [reply salt/modules/cmdmod.py:605]: Why uses 437? Python is utf-8 so why not just use 65001 by default? Also the code page change does not take effect until the second call. So we can just call _run with just the code page change so it takes effect right away!

**Review by cmcmarrow — CHANGES_REQUESTED** [review]
Why use 437? Python is utf-8 so why not just use 65001 by default? Also the code page change does not take effect until the second call. So we can just call _run with just the code page change so it takes effect right away!

**Comment by cmcmarrow** [comment]
It alos looks like the `echo Datenträger` test is not effective

**Comment by cmcmarrow** [comment]
Also `chcp 65001 > nul` gives me an error but `chcp 65001 > null` does not

**Comment by markuskramerIgitt** [comment]
@chad, on the 3001 Minion I successfully hacked two things within cmdmod.py:
1) I removed `chcp 437 > nul &`
2) I added `subprocess.run("C:\\Windows\\System32\\chcp.com 65001", stdout=subprocess.DEVNULL)` before  `proc = salt.utils.timed_subprocess.TimedProc(cmd, **new_kwargs)`


Success means letter ä is decoded always (and from the first time)

hack means I do not want to make a pull request with that code. 

I don't understand where you call _run(), but hopefully your and my solution are more or less doing the same,



These are lines 687: 

```
    if not use_vt:
        # This is where the unicode magic happens
        try:
            subprocess.run("C:\\Windows\\System32\\chcp.com 65001", stdout=subprocess.DEVNULL)
        except (BaseException) as exc:
            msg = (
                "Unable to run chcp 65001 before command '{0}' , "
                "reason: {1}".format(
                    cmd if output_loglevel is not None else "REDACTED", exc
                )
            )
            raise CommandExecutionError(msg)


        # This is where the magic happens
```


image

**Pull request closed without merging**